make.paws/R/xmlutil.R

```diff
@@ -1,16 +1,188 @@
-# Unmarshal `data` provided as XML into the shape in `interface`.
+# Decode raw bytes XML into an R list object.
+decode_xml <- function(raw) {
+  obj <- xml_to_list(rawToChar(raw))
+  return(obj)
+}
+
+# Convert an XML string to an R list.
+xml_to_list <- function(value) {
+  if (is_empty(value)) return(NULL)
+  result <- xml2::as_list(xml2::read_xml(value))
+  return(result)
+}
+
+add_xmlns <- function(xml_list, xmlns = "") {
+  if (!is.list(xml_list)) {
+    return(xml_list)
+  }
+  result <- lapply(xml_list, function(x) structure(add_xmlns(x, xmlns), xmlns = xmlns))
+  return(result)
+}
+
+# Convert list to XML text
+list_to_xml <- function(value) {
+  value_xml <- xml2::as_xml_document(x = value)
+  value_character <- as.character(value_xml, options = "no_declaration")
+  value_character <- gsub("\\n", "", value_character)
+  return(value_character)
+}
+
+#-------------------------------------------------------------------------------
+
+xml_build_body <- function(request) {
+  params <- request$params
+
+  if (is_empty(params)) {
+    body_xml <- ""
+    request$body <- body_xml
+    request$http_request$body <- body_xml
+    return(request)
+  }
+
+  locationName <- get_tag(params, "locationName")
+  xmlns <- get_tag(params, "xmlURI")
+  if (locationName != "") {
+    params <- Structure(params, .attrs = list(locationName = locationName,
+                                              xmlns = xmlns))
+  }
+  
+  body_list <- xml_build(params)
+
+  if (length(body_list)) {
+    if (xmlns != "") body_list <- add_xmlns(body_list, xmlns)
+    body_xml <- list_to_xml(body_list)
+  } else {
+    body_xml <- ""
+  }
+
+  request$body <- body_xml
+  request$http_request$body <- body_xml
+  return(request)
+}
+
+xml_build <- function(params) {
+
+  location <- get_tag(params, "location")
+  if (location != "") return(NULL)
+
+  t <- type(params)
+
+  build_fn <- switch(
+    t,
+    structure = xml_build_structure,
+    list = xml_build_list,
+    xml_build_scalar
+  )
+  
+  result <- build_fn(params)
+
+  return(result)
+}
+
+xml_build_structure <- function(params) {
+  result <- list()
+  if (is.null(names(params))) return(xml_build_structure_no_names(params))
+  for (name in names(params)) {
+    child <- params[[name]]
+
+    if (get_tag(child, "locationName") == "") {
+      child <- add_tags(list(locationName = name), child)
+    }
+
+    parsed <- xml_build(child)
+
+    if (!is.null(parsed)) {
+      location_name <- get_tag(child, "locationName")
+      if (location_name == "") location_name <- name
+
+      flattened <- get_tag(child, "flattened") != ""
+
+      if (flattened) {
+        result <- c(result, parsed)
+      } else{
+        result[[location_name]] <- parsed
+      }
+    }
+  }
+  return(result)
+}
+
+xml_build_structure_no_names <- function(params) {
```

Comment:
When does a structure have no names? To my knowledge it should always have names.